src/natcap/invest/delineateit/delineateit_core.pyx

```diff
@@ -0,0 +1,85 @@
+                                   int[:] edges, 
+                                   int nodata):
+    """
+    Return a binary array indicating which pixels are pour points.
+
+    Args:
+        flow_dir_array (numpy.ndarray): a 2D array of D8 flow direction 
+            values (0 - 7) and possiby the nodata value.
+        edges (dict): has boolean keys 'top', 'left', 'bottom', 'right'
+            indicating whether or not each edge is an edge of the raster.
+the 
+    Returns:
+        numpy.ndarray of the same shape as ``flow_direction_array``.
+        Each element is 1 if that pixel is a pour point, 0 if not,
+        nodata if it cannot be determined.
+    """   
+
+    # flow direction: (delta rows, delta columns)
+    # for a pixel at (i, j) with flow direction x,
+    # the pixel it flows into is located at:
+    # (i + directions[x][0], j + directions[x][1])
+    cdef int *row_offsets = [0, -1, -1, -1,  0,  1, 1, 1]
+    cdef int *col_offsets = [1,  1,  0, -1, -1, -1, 0, 1]
+
+    cdef int height, width
+    cdef int row, col
+    cdef int flow_dir
+    cdef int sink_row, sink_col
+
+    # tuple unpacking doesn't work in cython
+    height, width = flow_dir_array.shape[0], flow_dir_array.shape[1]
+
+    cdef cset[cpair[int, int]] pour_points
+
+    # iterate over each pixel
+    for row in range(height):
+        for col in range(width):
+            
+            # get the flow direction [0-7] for this pixel
+            flow_dir = flow_dir_array[row, col]
+            if flow_dir != nodata:
+                # get the location of the pixel it flows into
+                sink_row = row + row_offsets[flow_dir]
+                sink_col = col + col_offsets[flow_dir]
+
+                # if the row index is -1, the pixel is flowing off of the block
+                if sink_row == -1:
+                    # if this edge of the block is an edge of the raster, 
+                    # this is a pour point. Otherwise, we can't say whether it is
+                    # because the necessary information isn't in this block.
+                    if edges[0]:  # top edge
+                        pour_points.insert(cpair[int, int](col, row))
+                elif sink_col == -1:
+                    if edges[1]:  # left edge
+                        pour_points.insert(cpair[int, int](col, row))
+                elif sink_row == height:
+                    if edges[2]:  # bottom edge
+                        pour_points.insert(cpair[int, int](col, row))
+                elif sink_col == width:
+                    if edges[3]:  # right edge
+                        pour_points.insert(cpair[int, int](col, row))
+
+                # if we get to here, the point (sink_row, sink_col)  
+                # is known to be within the bounds of the array,
+                # so it's safe to index 
+                elif flow_dir_array[sink_row, sink_col] == nodata:
+                    pour_points.insert(cpair[int, int](col, row))
```

Comment:
Ah, I remember us talking a little about the possible need to restate things in cython, and I now see what you were referring to.  What you have here should work great!

If you would like to write this in a more abbreviated form, I'd suggest rolling the edge-checking into a single conditional which, [according to the python spec](https://docs.python.org/3/library/stdtypes.html#boolean-operations-and-or-not) (and also the C spec, not that I have a link to prove it), will use [short-circuit evaluation](https://en.wikipedia.org/wiki/Short-circuit_evaluation):

```python
if ((sink_row == -1 and edges[0]) or  # top edge
        (sink_col == -1 and edges[1]) or  # left edge
        (sink_row == height and edges[2]) or  # bottom edge
        (sink_col == width and edges[3]) or # right edge
        (flow_dir_array[sink_row, sink_col] == nodata)):  # internal pixel
    pour_points.insert(cpair[int, int](col, row))
```

Happy to talk about this more if you like, but just wanted to offer this as an alternative possible solution.